### DISCUSSION TITLE
force reloading/fetching in child components

### DISCUSSION THREAD

**Discussion by niklasrammelkamp** [neuefische/questions#514] [Web] [answered]
Hi there,
I have a component that handles a like function. So it's a Button with a heart icon and the functionality for handling the onClick event.
So far the component gets three values. The Post, the user ID of the logged-in user, and the mutate function for refetching. 'Cause by clicking the like button it is fetching and adds the user ID to the post and the post ID to the user. So when I click I have to refetch 'cause otherwise, there won't be seen any difference.

Sorry for that long intro but my question now is: Is possible to refetch without passing on the mutate function through several components? Or maybe you have a whole different idea for that component without passing on all three values, then let me know too.

Please have a look at my [Pull Request](https://github.com/niklasrammelkamp/capstone-project/pull/29#issue-1574811534) for that functionality. I'm talking about the ButtonLike component.

Thanks for your help and let me know if you need any further information.

**Comment by NikRoe** [comment, marked as answer]
Hi Niklas,

you can get the mutate function directly in your button component and revalidate your data by using the useSWR hook.

It can look similar to this:
```js
const { mutate } = useSWR(`/api/posts/${post._id}`);
```

instead of your `reload` function you can then call `mutate`.

Hope this helps :)

**Reply by niklasrammelkamp** [reply]
Nice one. Thank you, but unfortunately not perfect. The Thing is, that I'm using the Button on my post details page and on the feed page where I fetch all my postings and not just a single one. 
On the details page, it's working absolutely fine but on the feed, it's not fetching again the whole feed.

**Reply by NikRoe** [reply]
Ah, I see. 

We rediscussed your issue and the easiest thing would be to leave your code as it is.
Passing down the `mutate` function as `reload` is actually a nice solution.

For my proposed solution to work properly you would need useSWR with two different keys and you would need to decide upon which key will be passed to the hook. This would most likely be done on a higher level and this would make the solution worse than what you currently have. (Since we would depend on the URL we pass in as a prop to be correct)

The best solution, but also by far the most time intensive, would be to refactor your LikeButton component as a whole.
Your LikeButton should be as dumb as possible, meaning there is next to no logic being handled. This would mean that your functions for liking and disliking a post would no longer be part of that component, but would live somewhere higher up. They would only be passed down as a prop to your LikeButton and be called when that button is clicked.
All of your logic would be handled in that handler function (included the mutate function).

**Reply by niklasrammelkamp** [reply]
Nice thank you, I had it like that before with a dumb component and I like both. I think I leave it like that 'cause now I only need to add the component wherever I want and it works.

Thank you :)
